### ISSUE TITLE
Disable constant propagation per rule was: False positive in python.lang.correctness.useless-comparison.no-strings-as-booleans

### ISSUE THREAD

**Issue by moshekaplan** [returntocorp/sgrep-rules#1134]
**Describe the bug**
`if var:`  where `var` is a string is not a mistake - that is common behavior, equivalent to checking `if len(var) > 0:`. It would only be a bug if the string is hardcoded, as that is no different than `if True:`  or `if False:`

**To Reproduce**
The following code was triggered as a problem:
`if (opt): opt += '|'`

This line of code was seen in: https://github.com/wireshark/wireshark/blob/master/tools/asn2wrs.py

A minimal test case is available here: https://semgrep.dev/s/kX12

```python
def foo(variable):
    opt = ''
    if (variable):
        opt = 'bar'
    if (opt):
        opt += 'baz'
    return opt
```

**Expected behavior**
Do not mark `if (str)` as a problem

**Additional Context**
https://gitlab.com/wireshark/wireshark/-/merge_requests/1797

**Comment by underyx** [comment]
We should likely change this indeed. I assume this rule got a lot more sensitive when we released constant propagation awareness, and we just never realized. Although I will mention that in the sample code you linked:

```python
opt = ''
if (optional):
    opt = 'BER_FLAGS_OPTIONAL'
if (not val.HasOwnTag()):
    if (opt): opt += '|'
    opt += 'BER_FLAGS_NOOWNTAG'
elif (val.HasImplicitTag(ectx)):
    if (opt): opt += '|'
    opt += 'BER_FLAGS_IMPLTAG'
if (val.IndetermTag(ectx)):
    if (opt): opt += '|'
    opt += 'BER_FLAGS_NOTCHKTAG'
if (not opt): opt = '0'
```

Semgrep takes issue not with a string type being used in the 'if' condition, but with the fact that `opt` in the condition is always a constant string.

**Comment by mschwager** [comment]
> I assume this rule got a lot more sensitive when we released constant propagation awareness ...

Ahh, it all makes sense now! I'll take a look at this tomorrow :+1:

**Comment by mschwager** [comment]
I see now, I was misunderstanding where the findings were. I created this snippet to test: https://semgrep.dev/s/DoYo/.

This looks like a bug in our [data-flow analysis](https://semgrep.dev/docs/experiments/overview/#data-flow-analysis). I would expect it to register that `opt` can be truthy from `opt = 'BER_FLAGS_OPTIONAL'`. I'll pull in our resident data-flow expert, @IagoAbal. Can we catch this case?

**Comment by underyx** [comment]
@mschwager I don't think this is a data-flow bug. The `"..."` in the pattern doesn't care if the string is truthy or falsey. See demonstration here: https://semgrep.dev/s/0QZ5/

**Comment by mschwager** [comment]
Hmm, if I'm understanding the rule correctly, I think both of your findings are correct. Since both are either always truthy, or always falsey, then the `if` condition is useless. I think with dataflow we've lost the ability to express the difference between the following:

```
$VAR = "..."
...
if $VAR:
    ...
```

```
if "...":
    ...
```

The second example is searching for a constant string, whether `"stuff"` or `""`, because that will always be truthy or falsey. I think `if "..."` has effectively become "is this is a string data type?" vs. "is this a constant string?" This isn't necessarily a bad thing, I think we've just lost the ability to effectively express this rule. I see a few options, we could: 1.) remove the rule - I suspect this isn't one of our most impactful rules, or 2.) add the ability to express these two different types of string considerations. Or, is there a way we can improve the rule to remove these FPs?

**Comment by IagoAbal** [comment]
>  I think with dataflow we've lost the ability to express the difference between the following

Essentially yes.

> I think if "..." has effectively become "is this is a string data type?"

That is not entirely correct.

Pattern  `"..."` not only stands for a string literal, it stands for any string expression whose value is made of compile-time constants only. If a string expression depends on e.g. an user input then it will not match `"..."`. Note that we don't need to know the exact value of the string expression, just knowing that it is e.g. either "abc" or "def" would be enough for it to match `"..."`.

IMO we should have a construct in the rule language to choose between different degrees of constant propagation. This would be an easy feature to add on my side.

**Comment by minusworld** [comment]
@IagoAbal That would be swell! If you are able to set it up, I can create a way to control it in the rule language.

@minusworld Note to self: we'll need to document this after implementation.

(I'm marking this as low priority unless the OP needs this functionality soon.)

**Comment by moshekaplan** [comment]
As a quick note: I added a shorter clearer test case to the issue description. Specifically:

https://semgrep.dev/s/kX12

```python
def foo(variable):
    opt = ''
    if (variable):
        opt = 'bar'
    if (opt):
        opt += 'baz'
    return opt
```

Note that it flags on line 5 (`if (opt):`)

**Comment by aryx** [comment]
This might help at some point: https://github.com/returntocorp/semgrep/pull/2723

**Comment by IagoAbal** [comment]
@minusworld PR #2723 by @aryx allows us to "disable" constant propagation at the pattern or rule level, so it basically remains adding the corresponding config in the rule language.

**Comment by minusworld** [comment]
Excellent. I'll take a look at wiring it up soon. 🎉

**Comment by aryx** [comment]
@minusworld once you add the --experimental flag to pass the whole rule to semgrep, we can talk about adding a field in the yaml to enable/disable certain equivalences and you can just pass it and I'll use the information in semgrep-core.

**Comment by IagoAbal** [comment]
AFAICT this should be a non-issue now, we can just add to the rule:
```yaml
options:
  constant_propagation: false
```
